File: 01-ll-designs/parking_lot/solutions/go/parking_lot.go
package main

import (
	"fmt"
	"strings"
	"sync"
	"time"
)

// ParkingLot represents the main parking lot management system
type ParkingLot struct {
	mu            sync.RWMutex
	Name          string                    `json:"name"`
	Levels        []*ParkingLevel           `json:"levels"`
	PricingPolicy PricingPolicy             `json:"-"`
	ActiveTickets map[string]*Ticket        `json:"active_tickets"`
	SpotToLicense map[string]string         `json:"-"` // "level-spotId" -> licensePlate
}

// NewParkingLot creates a new parking lot
func NewParkingLot(name string, levels []*ParkingLevel) *ParkingLot {
	return &ParkingLot{
		Name:          strings.TrimSpace(name),
		Levels:        levels,
		PricingPolicy: NewStandardPricingPolicy(),
		ActiveTickets: make(map[string]*Ticket),
		SpotToLicense: make(map[string]string),
	}
}

// SetPricingPolicy sets a custom pricing policy
func (pl *ParkingLot) SetPricingPolicy(policy PricingPolicy) {
	pl.mu.Lock()
	defer pl.mu.Unlock()
	pl.PricingPolicy = policy
}

// ParkVehicle parks a vehicle and returns a ticket if successful
func (pl *ParkingLot) ParkVehicle(vehicle *Vehicle) (*Ticket, error) {
	if vehicle == nil {
		return nil, &ParkingError{Op: "park", Msg: "vehicle cannot be nil"}
	}
	
	pl.mu.Lock()
	defer pl.mu.Unlock()
	
	licensePlate := vehicle.LicensePlate
	
	// Check if vehicle is already parked
	if _, exists := pl.ActiveTickets[licensePlate]; exists {
		return nil, &ParkingError{
			Op:  "park",
			Msg: fmt.Sprintf("vehicle %s is already parked", licensePlate),
		}
	}
	
	// Try to find a spot across all levels
	for _, level := range pl.Levels {
		spotIndex, err := level.FindAvailableSpot(vehicle.Type)
		if err != nil {
			continue // Try next level
		}
		
		// Get the spot and occupy it
		spot, err := level.GetSpot(spotIndex)
		if err != nil {
			continue // This shouldn't happen, but handle gracefully
		}
		
		if err := spot.Occupy(licensePlate); err != nil {
			continue // Spot was taken by another goroutine, try next
		}
		
		// Create ticket
		spotID, spotType := spot.GetInfo()
		ticket := NewTicket(licensePlate, vehicle.Type, level.Index, spotID, spotType)
		
		// Update tracking maps
		pl.ActiveTickets[licensePlate] = ticket
		pl.SpotToLicense[pl.getSpotKey(level.Index, spotID)] = licensePlate
		
		return ticket, nil
	}
	
	return nil, ErrNoAvailableSpots
}

// UnparkVehicle unparks a vehicle and returns the fee charged
func (pl *ParkingLot) UnparkVehicle(ticket *Ticket) (float64, error) {
	if ticket == nil {
		return 0, &ParkingError{Op: "unpark", Msg: "ticket cannot be nil"}
	}
	
	pl.mu.Lock()
	defer pl.mu.Unlock()
	
	licensePlate := ticket.LicensePlate
	
	// Verify ticket is valid
	storedTicket, exists := pl.ActiveTickets[licensePlate]
	if !exists {
		return 0, &ParkingError{
			Op:  "unpark",
			Msg: fmt.Sprintf("ticket for %s not found in active tickets", licensePlate),
		}
	}
	
	if storedTicket.ID != ticket.ID {
		return 0, &ParkingError{
			Op:  "unpark",
			Msg: fmt.Sprintf("ticket mismatch for %s", licensePlate),
		}
	}
	
	// Find the level and spot
	level := pl.findLevel(ticket.LevelIndex)
	if level == nil {
		return 0, &ParkingError{
			Op:  "unpark",
			Msg: fmt.Sprintf("level %d not found", ticket.LevelIndex),
		}
	}
	
	spotIndex := level.FindSpotIndexByID(ticket.SpotID)
	if spotIndex == -1 {
		return 0, &ParkingError{
			Op:  "unpark",
			Msg: fmt.Sprintf("spot %d not found in level %d", ticket.SpotID, ticket.LevelIndex),
		}
	}
	
	spot, err := level.GetSpot(spotIndex)
	if err != nil {
		return 0, err
	}
	
	// Verify spot occupancy
	isOccupied, currentLicense := spot.GetStatus()
	if !isOccupied || currentLicense != licensePlate {
		return 0, &ParkingError{
			Op:  "unpark",
			Msg: fmt.Sprintf("spot occupancy mismatch for %s", licensePlate),
		}
	}
	
	// Release the spot
	if err := level.ReleaseSpot(spotIndex); err != nil {
		return 0, err
	}
	
	// Calculate fee
	exitTime := time.Now()
	fee := pl.PricingPolicy.CalculateFee(ticket.VehicleType, ticket.EntryTime, exitTime)
	
	// Clean up tracking maps
	delete(pl.ActiveTickets, licensePlate)
	delete(pl.SpotToLicense, pl.getSpotKey(ticket.LevelIndex, ticket.SpotID))
	
	return fee, nil
}

// GetAvailabilitySummary returns a formatted string with availability information
func (pl *ParkingLot) GetAvailabilitySummary() string {
	pl.mu.RLock()
	defer pl.mu.RUnlock()
	
	var summary strings.Builder
	summary.WriteString(fmt.Sprintf("Parking Lot: %s\n", pl.Name))
	
	totalMotorcycle, totalCompact, totalLarge := 0, 0, 0
	
	for _, level := range pl.Levels {
		motorcycle, compact, large := level.GetAvailability()
		totalMotorcycle += motorcycle
		totalCompact += compact
		totalLarge += large
		summary.WriteString(fmt.Sprintf("%s\n", level.String()))
	}
	
	summary.WriteString(fmt.Sprintf("Total available: %d motorcycle, %d compact, %d large",
		totalMotorcycle, totalCompact, totalLarge))
	
	return summary.String()
}

// GetActiveTickets returns a copy of all active tickets
func (pl *ParkingLot) GetActiveTickets() []*Ticket {
	pl.mu.RLock()
	defer pl.mu.RUnlock()
	
	tickets := make([]*Ticket, 0, len(pl.ActiveTickets))
	for _, ticket := range pl.ActiveTickets {
		tickets = append(tickets, ticket)
	}
	return tickets
}

// IsVehicleParked checks if a vehicle is currently parked
func (pl *ParkingLot) IsVehicleParked(licensePlate string) bool {
	pl.mu.RLock()
	defer pl.mu.RUnlock()
	
	_, exists := pl.ActiveTickets[licensePlate]
	return exists
}

// GetTicketForVehicle returns the ticket for a parked vehicle
func (pl *ParkingLot) GetTicketForVehicle(licensePlate string) *Ticket {
	pl.mu.RLock()
	defer pl.mu.RUnlock()
	
	return pl.ActiveTickets[licensePlate]
}

// GetName returns the parking lot name
func (pl *ParkingLot) GetName() string {
	pl.mu.RLock()
	defer pl.mu.RUnlock()
	return pl.Name
}

// GetLevels returns a copy of all levels
func (pl *ParkingLot) GetLevels() []*ParkingLevel {
	pl.mu.RLock()
	defer pl.mu.RUnlock()
	
	levels := make([]*ParkingLevel, len(pl.Levels))
	copy(levels, pl.Levels)
	return levels
}

// getSpotKey generates a spot key for mapping
func (pl *ParkingLot) getSpotKey(levelIndex, spotID int) string {
	return fmt.Sprintf("%d-%d", levelIndex, spotID)
}

// findLevel finds a level by index (must be called with lock held)
func (pl *ParkingLot) findLevel(levelIndex int) *ParkingLevel {
	for _, level := range pl.Levels {
		if level.Index == levelIndex {
			return level
		}
	}
	return nil
}

func (pl *ParkingLot) String() string {
	return pl.GetAvailabilitySummary()
}

File: 01-ll-designs/parking_lot/solutions/java/ParkingLot.java
import java.util.*;

/**
 * Main parking lot management system.
 * Coordinates multiple levels and handles vehicle parking/unparking operations.
 */
public class ParkingLot {
    private final String name;
    private final List<ParkingLevel> levels;
    private final PricingPolicy pricingPolicy;
    
    // Maps for quick lookups
    private final Map<String, Ticket> activeTickets; // licensePlate -> ticket
    private final Map<String, String> spotToLicense; // "level-spotId" -> licensePlate
    
    public ParkingLot(String name, List<ParkingLevel> levels) {
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("Parking lot name cannot be null or empty");
        }
        if (levels == null || levels.isEmpty()) {
            throw new IllegalArgumentException("Parking lot must have at least one level");
        }
        
        this.name = name.trim();
        this.levels = new ArrayList<>(levels);
        this.pricingPolicy = new PricingPolicy();
        this.activeTickets = new HashMap<>();
        this.spotToLicense = new HashMap<>();
    }
    
    /**
     * Park a vehicle and return a ticket if successful.
     * @param vehicle Vehicle to park
     * @return Ticket if parking successful, null otherwise
     */
    public Ticket parkVehicle(Vehicle vehicle) {
        if (vehicle == null) {
            throw new IllegalArgumentException("Vehicle cannot be null");
        }
        
        String licensePlate = vehicle.getLicensePlate();
        
        // Check if vehicle is already parked
        if (activeTickets.containsKey(licensePlate)) {
            System.out.println("Vehicle " + licensePlate + " is already parked!");
            return null;
        }
        
        // Try to find a spot across all levels
        for (ParkingLevel level : levels) {
            int spotIndex = level.findAvailableSpot(vehicle.getVehicleType());
            if (spotIndex != -1) {
                // Allocate the spot
                ParkingSpot spot = level.getSpot(spotIndex);
                spot.occupy(licensePlate);
                
                // Create ticket
                Ticket ticket = new Ticket(
                    generateTicketId(licensePlate, level.getLevelIndex(), spot.getSpotId()),
                    licensePlate,
                    vehicle.getVehicleType(),
                    System.currentTimeMillis(),
                    level.getLevelIndex(),
                    spot.getSpotId(),
                    spot.getSpotType()
                );
                
                // Update tracking maps
                activeTickets.put(licensePlate, ticket);
                spotToLicense.put(getSpotKey(level.getLevelIndex(), spot.getSpotId()), licensePlate);
                
                return ticket;
            }
        }
        
        // No available spots
        return null;
    }
    
    /**
     * Unpark a vehicle and return the fee charged.
     * @param ticket Parking ticket
     * @return Fee charged, or -1 if unparking failed
     */
    public double unparkVehicle(Ticket ticket) {
        if (ticket == null) {
            throw new IllegalArgumentException("Ticket cannot be null");
        }
        
        String licensePlate = ticket.getLicensePlate();
        
        // Verify ticket is valid
        if (!activeTickets.containsKey(licensePlate)) {
            System.out.println("Invalid ticket: " + licensePlate + " not found in active tickets");
            return -1;
        }
        
        Ticket storedTicket = activeTickets.get(licensePlate);
        if (!storedTicket.getTicketId().equals(ticket.getTicketId())) {
            System.out.println("Ticket mismatch for " + licensePlate);
            return -1;
        }
        
        // Find the level and spot
        ParkingLevel level = findLevel(ticket.getLevelIndex());
        if (level == null) {
            System.out.println("Level " + ticket.getLevelIndex() + " not found");
            return -1;
        }
        
        int spotIndex = level.findSpotIndexById(ticket.getSpotId());
        if (spotIndex == -1) {
            System.out.println("Spot " + ticket.getSpotId() + " not found in level " + ticket.getLevelIndex());
            return -1;
        }
        
        ParkingSpot spot = level.getSpot(spotIndex);
        if (!licensePlate.equals(spot.getCurrentVehicleLicense())) {
            System.out.println("Spot occupancy mismatch for " + licensePlate);
            return -1;
        }
        
        // Release the spot
        level.releaseSpot(spotIndex);
        
        // Calculate fee
        long exitTime = System.currentTimeMillis();
        double fee = pricingPolicy.calculateFee(ticket.getVehicleType(), ticket.getEntryTime(), exitTime);
        
        // Clean up tracking maps
        activeTickets.remove(licensePlate);
        spotToLicense.remove(getSpotKey(ticket.getLevelIndex(), ticket.getSpotId()));
        
        return fee;
    }
    
    /**
     * Get availability summary for all levels.
     * @return Formatted string with availability information
     */
    public String getAvailabilitySummary() {
        StringBuilder summary = new StringBuilder();
        summary.append("Parking Lot: ").append(name).append("\n");
        
        int totalMotorcycle = 0, totalCompact = 0, totalLarge = 0;
        
        for (ParkingLevel level : levels) {
            int[] availability = level.getAvailability();
            totalMotorcycle += availability[0];
            totalCompact += availability[1];
            totalLarge += availability[2];
            summary.append(level.toString()).append("\n");
        }
        
        summary.append(String.format("Total available: %d motorcycle, %d compact, %d large",
                totalMotorcycle, totalCompact, totalLarge));
        
        return summary.toString();
    }
    
    /**
     * Get list of all active tickets.
     * @return Collection of active tickets
     */
    public Collection<Ticket> getActiveTickets() {
        return new ArrayList<>(activeTickets.values());
    }
    
    /**
     * Check if a vehicle is currently parked.
     * @param licensePlate License plate to check
     * @return true if vehicle is parked, false otherwise
     */
    public boolean isVehicleParked(String licensePlate) {
        return activeTickets.containsKey(licensePlate);
    }
    
    /**
     * Get ticket for a parked vehicle.
     * @param licensePlate License plate
     * @return Ticket if vehicle is parked, null otherwise
     */
    public Ticket getTicketForVehicle(String licensePlate) {
        return activeTickets.get(licensePlate);
    }
    
    public String getName() {
        return name;
    }
    
    public List<ParkingLevel> getLevels() {
        return new ArrayList<>(levels);
    }
    
    /**
     * Generate a unique ticket ID.
     * @param licensePlate Vehicle license plate
     * @param levelIndex Level index
     * @param spotId Spot ID
     * @return Unique ticket ID
     */
    private String generateTicketId(String licensePlate, int levelIndex, int spotId) {
        long timestamp = System.currentTimeMillis();
        return String.format("%s-L%d-S%d-%d", licensePlate, levelIndex, spotId, timestamp);
    }
    
    /**
     * Generate spot key for mapping.
     * @param levelIndex Level index
     * @param spotId Spot ID
     * @return Spot key string
     */
    private String getSpotKey(int levelIndex, int spotId) {
        return levelIndex + "-" + spotId;
    }
    
    /**
     * Find level by index.
     * @param levelIndex Index to find
     * @return ParkingLevel or null if not found
     */
    private ParkingLevel findLevel(int levelIndex) {
        for (ParkingLevel level : levels) {
            if (level.getLevelIndex() == levelIndex) {
                return level;
            }
        }
        return null;
    }
    
    @Override
    public String toString() {
        return getAvailabilitySummary();
    }
}